---
title: "IDR Testing"
author: "Sarah Arcos"
date: "1/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(RCAS)
library(readxl)
BiocManager::install("ChIPseeker")

library(BSgenome.Hsapiens.UCSC.hg19)
```

#Fix naming of gene_biotype column in gtf rds
```{r}
# gtf <- readRDS("~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.granges.rds")
# 
# gtf$gene_biotype <- gtf$gene_type
# gtf$gene_type <- NULL
# 
# saveRDS(gtf, "~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.granges.rds")

```


# Read in data
```{r}
r2_bw3 <- readRDS("~/files/PARCLIP_IDR/data/df2_rep2.rda")
r3_bw3 <- readRDS("~/files/PARCLIP_IDR/data/df2_rep3.rda")

r_fracs <- readRDS("~/files/PARCLIP_IDR/data/r2_r3_overlaps.rda")

idr_mc <- readRDS("~/files/PARCLIP_IDR/data/idr_summ_T2Cfraction_MC1000")

m6A_r1 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep1.rda")
m6A_r2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep2.rda")
m6A_r1_q0.2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep1_q0.2.rda")
m6A_r2_q0.2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep2_q0.2.rda")

m6A_overlaps <- readRDS("~/files/PARCLIP_IDR/data/m6A_overlaps.rda")

idr_m6A <- readRDS("~/files/PARCLIP_IDR/data/idr_summ_m6A_MC1000")


df2_m6A <- readRDS("~/files/PARCLIP_IDR/data/df2_m6A.rda")
```

## Make everything in the same consistent format
 - GRanges
```{r}
df2_rep2 <- GRanges(seqnames = r2_bw3$Chromosome, IRanges(start = r2_bw3$Start, end = r2_bw3$End), strand = r2_bw3$Strand) %>%
  keepStandardChromosomes(pruning.mode="coarse")
df2_rep3 <- GRanges(seqnames = r3_bw3$Chromosome, IRanges(start = r3_bw3$Start, end = r3_bw3$End), strand = r3_bw3$Strand) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep1 <- GRanges(seqnames = m6A_r1$X1, IRanges(start = m6A_r1$X2, end = m6A_r1$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep2 <- GRanges(seqnames = m6A_r2$X1, IRanges(start = m6A_r2$X2, end = m6A_r2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")

df2_ov <- findOverlaps(df2_rep2, df2_rep3) %>%
  as_tibble()

df2 <- GRanges(seqnames = r3_bw3$Chromosome[df2_ov$subjectHits], IRanges(start = r3_bw3$Start[df2_ov$subjectHits], end = r3_bw3$End[df2_ov$subjectHits])) %>%
  keepStandardChromosomes(pruning.mode="coarse")

m6A_ov <- findOverlaps(m6A_rep1, m6A_rep2) %>%
  as_tibble()

m6A <- GRanges(seqnames = m6A_r1$X1[m6A_ov$queryHits], IRanges(start = m6A_r1$X2[m6A_ov$queryHits], end = m6A_r1$X3[m6A_ov$queryHits])) %>%
  keepStandardChromosomes(pruning.mode="coarse")

getSeq(Hsapiens, m6A)
```



#Motifs
```{r}
df2_m6A %>%
  filter(m6A == "m6A") %>%
  dplyr::select(Chromosome, Start, End, r3_name, T2Cfraction) %>%
  write_tsv("df2_m6A.bed", col_names = FALSE)

tb <- importBed('df2_m6A.bed')
tg <- importGtf('~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf')

motifResults <- runMotifRG(queryRegions = tb, 
                           resizeN = 15,
                           genomeVersion = 'hg19', 
                           motifN = 3, nCores = 8)

par(mfrow = c(1,2), mar = c(2,2,2,2))
for (i in 1:length(motifResults$motifs)) {
  motifPattern <- motifResults$motifs[[i]]@pattern
  motifRG::plotMotif(match = motifResults$motifs[[i]]@match$pattern, 
                     main = paste0('Motif-',i,': ',motifPattern),
                     entropy = TRUE)
}

df2_m6A %>%
  filter(m6A != "m6A") %>%
  dplyr::select(Chromosome, Start, End, r3_name, T2Cfraction) %>%
  write_tsv("df2_no_m6A.bed", col_names = FALSE)

tb_n <- importBed('df2_no_m6A.bed')

motifResults_n <- runMotifRG(queryRegions = tb_n, 
                           resizeN = 15,
                           genomeVersion = 'hg19', 
                           motifN = 3, nCores = 8)

par(mfrow = c(1,2), mar = c(2,2,2,2))
for (i in 1:length(motifResults_n$motifs)) {
  motifPattern_n <- motifResults_n$motifs[[i]]@pattern
  motifRG::plotMotif(match = motifResults_n$motifs[[i]]@match$pattern, 
                     main = paste0('Motif-',i,': ',motifPattern_n),
                     entropy = TRUE)
}
```

# Ribosome profiling test
 - Genes with no m6A peaks or DF2 peaks
 - Genes with m6A peaks but no DF2 peaks
 - Genes with DF2 peaks but no m6A peaks
 - Genes with both peaks and overlap
 - Genes with both peaks and no overlap
```{r}
ribo <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_C-Ribosome_profiling.xlsx", sheet = 3, skip = 3, 
                  col_names = c("Gene", "Ribosome-bound fragments", "Poly(A)+ mRNA input", "Translation Efficiency"))

ggplot(ribo, aes(x  = `Translation Efficiency`)) +
  stat_ecdf()

ribo %>%
  left_join(df2_m6A, by = c("Gene" = "Name")) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = m6A))

df2_peaks <- r3_bw3 %>%
  filter(Gene %in% r_fracs$r3_name)

m6A_peaks <- m6A_r1 %>%
  filter(X4 %in% m6A_overlaps$r1_name)

```

# runReport test
```{r}
tb <- importBed('df2_m6A.bed')
tg <- importGtf('~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf')

test <- queryGff(queryRegions = queryRegions, gff = gff)
overlaps.dt <- data.table(GenomicRanges::as.data.frame(test)) 
backgroundGenes <- unique(gff$gene_id)
targetedGenes <- unique(test$gene_id)

overlaps <- as.data.table(queryGff(queryRegions = tb, gffData = tg))
df <- overlaps[,length(unique(overlappingQuery)), by = gene_type]
colnames(df) <- c("feature", "count")
df$percent <- round(df$count / length(queryRegions) * 100, 1)
df <- df[order(count, decreasing = TRUE)]
ggplot2::ggplot(df, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = df$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(queryRegions), ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))


runReport('df2_m6A.bed', '~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf')
```




```{r}
sessionInfo()
```













