---
title: "IDR Testing"
author: "Sarah Arcos"
date: "1/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(tidyverse)
#library(RCAS)
library(readxl)
#devtools::install_github('BIMSBbioinfo/RCAS')
#BiocManager::install("org.Hs.eg.db")
#BiocManager::install("clusterProfiler")

library(GenomicFeatures)
library(clusterProfiler)
library(org.Hs.eg.db)
```

#Fix naming of gene_biotype column in gtf rds
```{r}
# gtf <- readRDS("~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.granges.rds")
# 
# gtf$gene_biotype <- gtf$gene_type
# gtf$gene_type <- NULL
# 
# saveRDS(gtf, "~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.granges.rds")
```


# Read in data
```{r}
r2_bw3 <- readRDS("~/files/PARCLIP_IDR/data/df2_rep2.rda")
r3_bw3 <- readRDS("~/files/PARCLIP_IDR/data/df2_rep3.rda")

#r_fracs <- readRDS("~/files/PARCLIP_IDR/data/r2_r3_overlaps.rda")

#idr_mc <- readRDS("~/files/PARCLIP_IDR/data/idr_summ_T2Cfraction_MC1000")

m6A_r1 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep1.rda")
m6A_r2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep2.rda")
m6A_r1_q0.2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep1_q0.2.rda")
m6A_r2_q0.2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep2_q0.2.rda")

#m6A_overlaps <- readRDS("~/files/PARCLIP_IDR/data/m6A_overlaps.rda")

#idr_m6A <- readRDS("~/files/PARCLIP_IDR/data/idr_summ_m6A_MC1000")


#df2_m6A <- readRDS("~/files/PARCLIP_IDR/data/df2_m6A.rda")
```

## Make everything in the same consistent format
 - GRanges
```{r}
df2_rep2 <- GRanges(seqnames = r2_bw3$Chromosome, IRanges(start = r2_bw3$Start, end = r2_bw3$End), strand = r2_bw3$Strand, name = r2_bw3$Name, score = r2_bw3$T2Cfraction) %>%
  keepStandardChromosomes(pruning.mode="coarse")
df2_rep3 <- GRanges(seqnames = r3_bw3$Chromosome, IRanges(start = r3_bw3$Start, end = r3_bw3$End), strand = r3_bw3$Strand, name = r3_bw3$Name, score = r3_bw3$T2Cfraction) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep1 <- GRanges(seqnames = m6A_r1$X1, IRanges(start = m6A_r1$X2, end = m6A_r1$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep2 <- GRanges(seqnames = m6A_r2$X1, IRanges(start = m6A_r2$X2, end = m6A_r2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")

df2 <- df2_rep2[df2_rep2 %over% df2_rep3,]
m6A <- m6A_rep2[m6A_rep2 %over% m6A_rep1,]

# df2_ov <- findOverlaps(df2_rep2, df2_rep3, maxgap = 1) %>%
#   as_tibble()
# 
# df2 <- GRanges(seqnames = r3_bw3$Chromosome[df2_ov$subjectHits], IRanges(start = r3_bw3$Start[df2_ov$subjectHits], end = r3_bw3$End[df2_ov$subjectHits]), strand = r3_bw3$Strand[df2_ov$subjectHits], name = r3_bw3$Name[df2_ov$subjectHits]) %>%
#   keepStandardChromosomes(pruning.mode="coarse")
# 
# m6A_ov <- findOverlaps(m6A_rep1, m6A_rep2, maxgap = 1) %>%
#   as_tibble()
# 
# m6A <- GRanges(seqnames = m6A_r1$X1[m6A_ov$queryHits], IRanges(start = m6A_r1$X2[m6A_ov$queryHits], end = m6A_r1$X3[m6A_ov$queryHits])) %>%
#   keepStandardChromosomes(pruning.mode="coarse")

```

#Look at low scoring m6A peaks to see if they overlap better
```{r}
m6A_rep1_q <- GRanges(seqnames = m6A_r1_q0.2$X1, IRanges(start = m6A_r1_q0.2$X2, end = m6A_r1_q0.2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep2_q <- GRanges(seqnames = m6A_r2_q0.2$X1, IRanges(start = m6A_r2_q0.2$X2, end = m6A_r2_q0.2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")

m6A_q <- m6A_rep2_q[m6A_rep2_q %over% m6A_rep1_q,]

# m6A_ov_q <- findOverlaps(m6A_rep1_q, m6A_rep2_q, maxgap = 1) %>%
#   as_tibble()
# 
# m6A_q <- GRanges(seqnames = m6A_r1_q0.2$X1[m6A_ov_q$queryHits], IRanges(start = m6A_r1_q0.2$X2[m6A_ov_q$queryHits], end = m6A_r1_q0.2$X3[m6A_ov_q$queryHits])) %>%
#   keepStandardChromosomes(pruning.mode="coarse")

df2_m6A_q <- df2[df2 %over% m6A_q,] %>%
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_"))

df2_m6A <- df2[df2 %over% m6A,] %>%
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_"))

df2 %>% as_tibble() %>% dim()

m6A %>% as_tibble() %>% dim()

m6A_q %>% as_tibble() %>% dim()


df2_non <- df2[!df2 %over% m6A,] %>% as_tibble()

df2_non_q <- df2[!df2 %over% m6A_q,] %>% as_tibble()

df2 %>% as_tibble() %>% dim()
df2_m6A %>% dim()
df2_non %>% dim()

df2 %>% as_tibble() %>% dim()
df2_m6A_q %>%  dim()
df2_non_q %>% dim()


m6A_df2_q <- m6A_q[m6A_q %over% df2,] %>%
  as_tibble()

m6A_df2 <- m6A[m6A %over% df2,] %>%
  as_tibble()

m6A_df2_q_non <- m6A_q[!m6A_q %over% df2,] %>%
  as_tibble()

m6A_df2_non <- m6A[!m6A %over% df2,] %>%
  as_tibble()

m6A %>% as_tibble() %>% dim()
m6A_df2 %>% as_tibble() %>% dim()
m6A_df2_non %>% as_tibble() %>% dim()


m6A_q %>% as_tibble() %>% dim()
m6A_df2_q %>% as_tibble() %>% dim()
m6A_df2_q_non %>% as_tibble() %>% dim()

```

#Make distribution of T2C by m6A status plot
```{r}
df2_annotated <- bind_rows(list("m6A" = df2_m6A, "non-m6A" = df2_non), .id = "m6A_status")

ggplot(df2_annotated, aes(y = score, x = m6A_status)) +
  geom_violin(fill = "steelblue", alpha = 0.7) +
  geom_boxplot(width = 0.03, alpha = 0.8, outlier.shape = NA) +
  labs(title = "Distribution of T2C Fraction in YTHDF peaks, based on m6A status") +
  theme_minimal() +
  theme(panel.grid.minor = element_blank()) +
  xlab("Modification state of YTHDF peaks") +
  annotate("text", x = 1.5, y = 0.9, label = "p = 3.8 * 10^-7")

wilcox.test(score ~ m6A_status, data = df2_annotated)

#Above T2Cfraction of 0.20

ggplot(df2_annotated %>% filter(score > 0.20), aes(y = score, x = m6A_status)) +
  geom_violin(fill = "steelblue", alpha = 0.7) +
  geom_boxplot(width = 0.03, alpha = 0.8, outlier.shape = NA) +
  labs(title = "Distribution of T2C Fraction in YTHDF peaks, based on m6A status") +
  theme_minimal() +
  theme(panel.grid.minor = element_blank()) +
  xlab("Modification state of YTHDF peaks") +
  annotate("text", x = 1.5, y = 0.9, label = "p = 0.006333")

wilcox.test(score ~ m6A_status, data = df2_annotated %>% filter(score > 0.20))

```



#Get tally for each gene of m6A sites, DF2 sites, and shared sites
 - First annotate DF2 sites by presence/absence of m6A
 - Then make dataframe of gene, DF2 site id, m6A site id, and p/a of m6A
 - Then make tallies
 - Then plot distribution

```{r}

df2_df <- df2 %>%
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_"))

df2_m6A_df <- df2_m6A
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_"))
  
all_peaks <- df2_annotated %>%
  group_by(name, m6A_status) %>%
  tally() %>%
  group_by(name) %>%
  mutate(total_peaks = sum(n)) %>%
  spread(key = m6A_status, value = n, fill = 0) %>%
  mutate(df2_only_frac = `non-m6A` / total_peaks)

ggplot(all_peaks, aes(x = df2_only_frac)) +
  geom_bar()

ggplot(all_peaks, aes(x = `non-m6A`)) +
  geom_bar()
```

#Look at whether m6A/non-m6A YTHDF2 peaks are enriched for certain protein domains
 - Using uniprot domains file downloaded from UCSC (hg19)
 - Come back and do by "functional" genes
```{r}
uniprot <- read_tsv("~/files/uniprot_domains_hg19.dms") %>%
  filter(!str_detect(dbName, "Unreviewed") &
           annotationType != "topological domain")

uniprot_g <- GRanges(seqnames = uniprot$`#chrom`, IRanges(start = uniprot$chromStart, end = uniprot$chromEnd), name = uniprot$name) %>%
  keepStandardChromosomes(pruning.mode="coarse")

df2_scored <- df2[df2$score > 0.2]

df2_o <- df2_scored[!df2_scored %over% m6A,]


df2_m <- df2_scored[df2_scored %over% m6A,]

domains_df2_o <- uniprot_g[uniprot_g %over% df2_o,] %>%
  as_tibble() %>%
  dplyr::select(name) %>%
  group_by(name) %>%
  tally() %>%
  mutate(freq = n / sum(n)) %>%
  top_n(5, freq)

domains_df2_m <- uniprot_g[uniprot_g %over% df2_m,] %>%
  as_tibble() %>%
  dplyr::select(name) %>%
  group_by(name) %>%
  tally() %>%
  mutate(freq = n / sum(n)) %>%
  top_n(5, freq)

domains_df2 <- bind_rows("non-m6A" = domains_df2_o, "m6A-modified" = domains_df2_m, .id = "m6A status")

ggplot2::ggplot(domains_df2, aes(x = reorder(name, n), y = n)) + 
  geom_bar(stat = 'identity') + 
  labs(x = 'Domain', y = "number overlapped with DF2 peaks") +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90)) +
  coord_flip() +
  facet_wrap(~`m6A status`, scales = "free_y")


gFeatures <- list(
  "zincfinger" = uniprot_g[uniprot_g$name == "zinc finger"],
  "proteinkinase" = uniprot_g[uniprot_g$name == "Protein kinase"])

cvB_non5 <- getFeatureBoundaryCoverage(
  queryRegions = tb,
  featureCoords = gFeatures$zincfinger,
  boundaryType = 'fiveprime'
)

cvB_non3 <- getFeatureBoundaryCoverage(
  queryRegions = tb,
  featureCoords = gFeatures$zincfinger,
  boundaryType = 'threeprime'
)

ggplot2::ggplot(cvB_non5, aes(x = bases, y = meanCoverage)) + 
  geom_ribbon(fill = 'lightblue', 
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
 geom_line(color = 'black')


```





#Motifs
```{r, eval = FALSE}
set.seed(5)


df2_m6A %>%
  filter(score > 0.2) %>%
  dplyr::select(seqnames, start, end, name, width, strand) %>%
  write_tsv("df2_m6A.bed", col_names = FALSE)

tb <- importBed('df2_m6A.bed')

motifResults <- runMotifDiscovery(queryRegions = tb, 
                           resizeN = 0, sampleN = 10000,
                           genomeVersion = 'hg19', motifWidth = 6,
                           motifN = 4, nCores = 4)

ggseqlogo::ggseqlogo(motifResults$matches_query)

df2_non %>%
  filter(score > 0.2) %>%
  dplyr::select(seqnames, start, end, name, width, strand) %>%
  write_tsv("df2_no_m6A.bed", col_names = FALSE)

tb_n <- importBed('df2_no_m6A.bed')

motifResults_n <- runMotifDiscovery(queryRegions = tb_n, 
                           resizeN = 0, sampleN = 10000,
                           genomeVersion = 'hg19', motifWidth = 6,
                           motifN = 4, nCores = 4)

ggseqlogo::ggseqlogo(motifResults_n$matches_query)
```

#Metagene
```{r}



gff <- importGtf(filePath = "~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf")
txdbFeatures <- getTxdbFeaturesFromGRanges(gff)

cvgList <- calculateCoverageProfileList(queryRegions = tb, 
                                        targetRegionsList = txdbFeatures, 
                                        sampleN = 0)

# ggplot2::ggplot(cvgList, aes(x = bins, y = meanCoverage)) + 
#   geom_ribbon(fill = 'lightgreen', 
#               aes(ymin = meanCoverage - standardError * 1.96, 
#                   ymax = meanCoverage + standardError * 1.96)) + 
#   geom_line(color = 'black') + theme_bw(base_size = 14) +
#   facet_wrap( ~ feature, ncol = 3) 

cvgList_tx <- cvgList %>%
  filter(feature  == "transcripts")

ggplot2::ggplot(cvgList_tx, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14)


cvgList_n <- calculateCoverageProfileList(queryRegions = tb_n, 
                                        targetRegionsList = txdbFeatures, 
                                        sampleN = 0)


cvgList_tx_n <- cvgList_n %>%
  filter(feature  == "transcripts")

ggplot2::ggplot(cvgList_tx_n, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14)

cvgList_all <- bind_rows(list("m6A-modified" = cvgList_tx, "Unmodified" = cvgList_tx_n), .id = "m6A")

ggplot2::ggplot(cvgList_all, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14) +
  facet_wrap( ~ m6A)

```


#Feature from ucsc gtf
```{r}

uniprot_s <- read_tsv("~/files/uniprot_structure_hg19.dms") %>%
  filter(!str_detect(dbName, "Unreviewed"))

structs <- GRanges(seqnames = uniprot_s$`#chrom`, IRanges(start = uniprot_s$chromStart, end = uniprot_s$chromEnd), type = uniprot_s$name) %>%
  keepStandardChromosomes(pruning.mode="coarse")

sFeatures <- list(
  "gff_beta" = structs[structs$type == "beta"],
  "gff_turn" = structs[structs$type == "turn"],
  "gff_helix" = structs[structs$type == "helix"],
  "gff_coil" = structs[structs$type == "coiled-coil"] )

cv_non <- calculateCoverageProfileList(
  queryRegions = tb_n,
  targetRegions = sFeatures
)

ggplot2::ggplot(cv_non, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14) +
  facet_wrap(~feature)

cv_m6A <- calculateCoverageProfileList(
  queryRegions = tb,
  targetRegions = sFeatures
)

ggplot2::ggplot(cv_m6A, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14) +
  facet_wrap(~feature)


cvB_non5 <- getFeatureBoundaryCoverage(
  queryRegions = tb_n,
  featureCoords = sFeatures$gff_helix,
  boundaryType = 'fiveprime'
)

cvB_non3 <- getFeatureBoundaryCoverage(
  queryRegions = tb_n,
  featureCoords = sFeatures$gff_helix,
  boundaryType = 'threeprime'
)

ggplot2::ggplot(cvB_non5, aes(x = bases, y = meanCoverage)) + 
  geom_ribbon(fill = 'lightblue', 
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
 geom_line(color = 'black')


```


#Gene regions
```{r}
summary <- summarizeQueryRegions(queryRegions = tb, 
                                 txdbFeatures = txdbFeatures)

df <- data.frame(summary)
df$feature <- rownames(df)
df_gene <- df %>%
  filter(feature %in% c("fiveUTRs", "cds", "threeUTRs")) %>%
  mutate(percent = round((count / sum(count)), 3) * 100)

df_exonintron <- df %>%
  filter(feature %in% c("exons", "introns")) %>%
  mutate(percent = round((count / sum(count)), 3) * 100)

ggplot2::ggplot(df_gene, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_gene$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb), ')')) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))

summary_n <- summarizeQueryRegions(queryRegions = tb_n, 
                                 txdbFeatures = txdbFeatures)

df_n <- data.frame(summary_n)
df_n$feature <- rownames(df_n)
df_n_gene <- df_n %>%
  filter(feature %in% c("fiveUTRs", "cds", "threeUTRs")) %>%
  mutate(percent = round((count / sum(count)), 3) * 100)

df_n_exonintron <- df_n %>%
  filter(feature %in% c("exons", "introns")) %>%
  mutate(percent = round((count / sum(count)), 3) * 100)

ggplot2::ggplot(df_n_gene, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_n_gene$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb_n), ')')) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))


df_all_genes <- bind_rows(list("m6A-modified" = df_gene, "Unmodified" = df_n_gene), .id = "m6A") %>%
  mutate(feature = factor(feature, levels = c("fiveUTRs", "cds", "threeUTRs")))

ggplot2::ggplot(df_all_genes, aes(x = m6A, y = percent, group = feature, fill = feature)) + 
  geom_bar(stat = 'identity', position = "stack") + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_all_genes$count) + 
  labs(x = 'transcript feature', y = "percent overlap") +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


df_all_exonintron <- bind_rows(list("m6A-modified" = df_exonintron, "Unmodified" = df_n_exonintron), .id = "m6A") %>%
  mutate(feature = factor(feature, levels = c("exons", "introns")))

ggplot2::ggplot(df_all_exonintron, aes(x = m6A, y = percent, group = feature, fill = feature)) + 
  geom_bar(stat = 'identity', position = "stack") + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_all_exonintron$count) + 
  labs(x = 'transcript feature', y = "percent overlap") +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


#Gene types
```{r}

overlaps <- as.data.table(queryGff(queryRegions = tb, gffData = gff))

biotype_col <- grep('gene_biotype', colnames(overlaps), value = T)
dfv <- overlaps[,length(unique(queryIndex)), by = biotype_col]
colnames(dfv) <- c("feature", "count")
dfv$percent <- round(dfv$count / length(tb) * 100, 1)
dfv <- dfv[order(count, decreasing = TRUE)]
dfv <- dfv %>% filter(feature %in% c("protein_coding", "miRNA", "lincRNA", "snoRNA", 
                                   "snRNA", "rRNA", "pseudogene"))
ggplot2::ggplot(dfv, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = dfv$count) + 
  scale_fill_brewer(palette = "Accent") +
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb), ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


overlaps_n <- as.data.table(queryGff(queryRegions = tb_n, gffData = gff))

dfv_n <- overlaps_n[,length(unique(queryIndex)), by = biotype_col]
colnames(dfv_n) <- c("feature", "count")
dfv_n$percent <- round(dfv_n$count / length(tb_n) * 100, 1)
dfv_n <- dfv_n[order(count, decreasing = TRUE)]
dfv_n <- dfv_n %>% filter(feature %in% c("protein_coding", "miRNA", "lincRNA", "snoRNA", 
                                   "snRNA", "rRNA", "pseudogene"))
ggplot2::ggplot(dfv_n, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = dfv_n$count) + 
  scale_fill_brewer(palette = "Accent") +
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb_n), ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

dfv_all <- bind_rows(list("m6A-modified" = dfv, "Unmodified" = dfv_n), .id = "m6A")


ggplot2::ggplot(dfv_all, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = dfv_all$count) + 
  scale_fill_brewer(palette = "Accent") +
  labs(x = 'gene type', y = 'percent overlap') + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none") +
  facet_wrap( ~ m6A)
```


#Ribosome profiling for publication
```{r}
ribo <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_C-Ribosome_profiling.xlsx", sheet = 3, skip = 3, 
                  col_names = c("Gene", "Ribosome-bound fragments", "Poly(A)+ mRNA input", "Translation Efficiency"))


all_peaks_filt <- df2_annotated %>%
  filter(score > 0.20) %>%
  group_by(name, m6A_status) %>%
  tally() %>%
  group_by(name) %>%
  mutate(total_peaks = sum(n)) %>%
  spread(key = m6A_status, value = n, fill = 0) %>%
  mutate(df2_only_frac = `non-m6A` / total_peaks)



ribo_groups <- ribo %>%
  left_join(all_peaks_filt, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac <= 0.5 ~ "Most peaks are m6A-modified",
    df2_only_frac > 0.5 ~ "Most peaks are unmodified",
    is.na(df2_only_frac) ~ "Non-target"
  ))

ribo_free <- ribo_groups %>%
  filter(category != "Most peaks are m6A-modified")

wilcox.test(`Translation Efficiency` ~ category, data = ribo_free)

ribo_shared <- ribo_groups %>%
  filter(category != "Most peaks are unmodified")

wilcox.test(`Translation Efficiency` ~ category, data = ribo_shared)

ribo_compare <- ribo_groups %>%
  filter(category != "Non-target")

wilcox.test(`Translation Efficiency` ~ category, data = ribo_compare)

ggplot(ribo_groups, aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5)) +
  annotate("text", x = -0.5, y = 0.92, label = "p = 2.176 * 10^-6") +
  annotate("text", x = -0.5, y = 0.85, label = "(Mostly m6A-modified vs.") +
  annotate("text", x = -0.5, y = 0.8, label = "Mostly Unmodified)") +
  ylab("Cumulative fraction") +
  xlab("Translation efficiency [log2(siYTHDF2/siControl)]") +
  labs(color = "Transcript category") +
  ggtitle("Translation efficiency of YTHDF2 targets by m6A-modification state") +
  theme_minimal() +
  theme(legend.position="bottom")

ribo_groups %>% 
  filter(category != "Non-target") %>%
  mutate(sig = case_when(
    `Translation Efficiency` < -2 ~ "Changed",
    TRUE ~ "un-changed"
  )) %>%
  ggplot(aes(x = category, y = `Translation Efficiency`, color = sig)) +
    geom_point(position = "jitter", alpha = 0.3) +
    geom_violin(color = "grey", fill = "grey", alpha = 0.5)
```

#Half life for publication
```{r}
hl_rep1 <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_E-mRNA_Lifetime-rep2.xlsx", sheet = 4, skip = 1) %>%
  dplyr::select(Gene, `log2(siYTHDF2/siControl)`)

hl_groups <- hl_rep1 %>%
  left_join(all_peaks_filt, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac <= 0.5 ~ "Most peaks are m6A-modified",
    df2_only_frac > 0.5 ~ "Most peaks are unmodified",
    is.na(df2_only_frac) ~ "Non-target"
  ))

hl_free <- hl_groups %>%
  filter(category != "Most peaks are m6A-modified")

wilcox.test(`log2(siYTHDF2/siControl)` ~ category, data = hl_free)

hl_shared <- hl_groups %>%
  filter(category != "Most peaks are unmodified")

wilcox.test(`log2(siYTHDF2/siControl)` ~ category, data = hl_shared)

hl_compare <- hl_groups %>%
  filter(category != "Non-target")

wilcox.test(`log2(siYTHDF2/siControl)` ~ category, data = hl_compare)

ggplot(hl_groups, aes(x = `log2(siYTHDF2/siControl)`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5)) +
  annotate("text", x = -0.5, y = 0.92, label = "p = 4.534 * 10^-7") +
  annotate("text", x = -0.5, y = 0.85, label = "(Mostly m6A-modified vs.") +
  annotate("text", x = -0.5, y = 0.8, label = "Mostly Unmodified)") +
  ylab("Cumulative fraction") +
  xlab("RNA half-life [log2(siYTHDF2/siControl)]") +
  labs(color = "Transcript category") +
  ggtitle("Half-life of YTHDF2 targets by m6A-modification state") +
  theme_minimal() +
  theme(legend.position="bottom")


hl_groups %>% 
  filter(category != "Non-target") %>%
  mutate(sig = case_when(
    `log2(siYTHDF2/siControl)` > 1 ~ "Changed",
    TRUE ~ "un-changed"
  )) %>%
  ggplot(aes(x = category, y = `log2(siYTHDF2/siControl)`, color = sig)) +
    geom_point(position = "jitter", alpha = 0.3) +
    geom_violin(color = "grey", fill = "grey", alpha = 0.5)
```

#Gene regions by ribo or stability
```{r}

ribo_changed <- ribo_groups %>%
  filter(`Translation Efficiency` < -0.5)

df2_non %>%
  filter(name %in% ribo_changed$Gene) %>%
  dplyr::select(seqnames, start, end, name, width, strand) %>%
  write_tsv("df2_m6A_ribo.bed", col_names = FALSE)

tb_rib <-  importBed('df2_m6A_ribo.bed')

summary <- summarizeQueryRegions(queryRegions = tb_rib, 
                                 txdbFeatures = txdbFeatures)

df <- data.frame(summary)
df$feature <- rownames(df)
df_gene <- df %>%
  filter(feature %in% c("fiveUTRs", "cds", "threeUTRs")) %>%
  mutate(percent = round((count / sum(count)), 3) * 100)

df_exonintron <- df %>%
  filter(feature %in% c("exons", "introns")) %>%
  mutate(percent = round((count / sum(count)), 3) * 100)

ggplot2::ggplot(df_gene, aes(x = 1, y = percent, group = feature, fill = feature)) + 
  geom_bar(stat = 'identity', position = "stack") + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_gene$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb_rib), ')')) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))

ggplot2::ggplot(df_exonintron, aes(x = 1, y = percent, group = feature, fill = feature)) + 
  geom_bar(stat = 'identity', position = "stack") + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_exonintron$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb_rib), ')')) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))
```



#Pulling out individual gene changes after DF2 KD
```{r}
ribo_r1 <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_C-Ribosome_profiling.xlsx", sheet = 1, skip = 2, 
                  col_names = c("Gene", "siCtr", "siDF2", NA, NA, NA, NA))

ribo_r2 <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_C-Ribosome_profiling.xlsx", sheet = 2, skip = 2, 
                  col_names = c("Gene", "siCtr", "siDF2", NA, NA, NA, NA, NA, NA, NA))

ribo_both <- inner_join(ribo_r1, ribo_r2, by = "Gene") %>%
  group_by(Gene) %>%
  mutate(siDF2_mean = mean(siDF2.x, siDF2.y, na.rm = TRUE),
         siCtr_mean = mean(siCtr.x, siCtr.y, na.rm = TRUE)) %>%
  select(Gene, siDF2_mean, siCtr_mean) %>%
  mutate(dif = log2(siDF2_mean / siCtr_mean)) %>%
  gather(Condition, `Ribosome protected fragments`, -Gene, -dif)

ggplot(ribo_both, aes(x = Condition, y = log2(`Ribosome protected fragments`), group = Gene)) +
  geom_line(alpha = 0.3)



ribo_groups_bygene <- ribo_both %>%
  left_join(all_peaks_filt, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac <= 0.5 ~ "Most peaks are m6A-modified",
    df2_only_frac > 0.5 ~ "Most peaks are unmodified",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  mutate(col = case_when(
    dif > 1 ~ "Increased translation",
    dif < -1 ~ "Decreased translation",
    TRUE ~ "Lame"
  ))

ggplot(ribo_groups_bygene,
       aes(x = Condition, y = log2(`Ribosome protected fragments`),
           group = Gene)) +
  geom_line(data = ribo_groups_bygene %>% filter(col == "Lame"),
            aes(x = Condition, y = log2(`Ribosome protected fragments`),
                group = Gene),
            alpha = 0.05, color = "grey") +
  geom_line(data = ribo_groups_bygene %>% filter(col == "Increased translation"),
            aes(x = Condition, y = log2(`Ribosome protected fragments`),
                group = Gene),
            alpha = 0.2, color = "red") +
  geom_line(data = ribo_groups_bygene %>% filter(col == "Decreased translation"),
            aes(x = Condition, y = log2(`Ribosome protected fragments`),
                group = Gene),
            alpha = 0.2, color = "blue") +
  facet_wrap(~category, ncol = 2) +
  theme_minimal()


ggplot(ribo_groups_bygene %>% filter(col != "Decreased translation" & category != "Non-target"),
       aes(x = Condition, y = log2(`Ribosome protected fragments`),
           group = Gene)) +
  geom_line(alpha = 0.05, color = "grey") +
  geom_line(data = ribo_groups_bygene %>% filter(dif < -2 & category != "Non-target"),
            aes(color = 1),
            alpha = 0.3) +
  facet_wrap(~category, ncol = 2) +
  theme_minimal()

```




```{r}
sessionInfo()
```













