---
title: "RCAS_Testing"
author: "Sarah Arcos"
date: "1/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# if (!requireNamespace("BiocManager", quietly=TRUE)) install.packages("BiocManager")
# BiocManager::install('RCAS')

# library(dplyr)
# library(readr)
# library(magrittr)
# library(tidyr)
# library(purrr)
# library(RColorBrewer)
# library(GenomicRanges)
# library(ggplot2)
# library(stringr)

#library(tidyverse)
library(idr)
#library(RCAS)
```

Below is old solution. Fixed problem by setting up my own docker image based on rocker/tidyverse:latest that installs RCAS and idr packages.
_So RCAS can't be installed because the R version is wrong, but I can't figure out how to update R on AWS. Temporary solution is to create necessary files, then scp them to my local machine._

```{r}
r2_bw3 <- readRDS("~/PARCLIP_IDR/data/df2_rep2.rda")
r3_bw3 <- readRDS("~/PARCLIP_IDR/data/df2_rep3.rda")

r_fracs <- readRDS("~/PARCLIP_IDR/data/r2_r3_overlaps.rda")

idr_mc <- readRDS("~/PARCLIP_IDR/data/idr_summ_T2Cfraction_MC1000")

m6A <- read_tsv("~/PARpipe/RCAS_files/m6A_peaks_hg19.bed", col_names = FALSE)
```

# Test overlaps with m6A dataset and distribution of scores
```{r}
#need to join idr results to original dfs to get peak start/stop info to calculate m6A overlaps. Use 3rd replicate
r3_idr <- idr_mc %>%
  left_join(r3_bw3, by = c("r3_name" = "Gene"))


m6A_g <- GRanges(m6A$X1, IRanges(m6A$X2, m6A$X3)) %>%
  keepStandardChromosomes(pruning.mode = "coarse")
df2_g <- GRanges(r3_idr$Chromosome, IRanges(r3_idr$Start, r3_idr$End), score = r3_idr$av, name = r3_idr$r3_name) %>%
  keepStandardChromosomes(pruning.mode="coarse")
  

ov_m6A <- as.data.frame(findOverlaps(m6A_g, df2_g))

df2_m6A <- as.data.frame(df2_g[ov_m6A$subjectHits])

r3_idr_m6A <- r3_idr %>%
  mutate(m6A = case_when(r3_name %in% df2_m6A$name ~ "m6A",
                         TRUE ~ "non-m6A"))
```

## Plot distributions
```{r}
ggplot(r3_idr_m6A, aes(x = av, group = m6A, color = m6A)) +
  geom_density() +
  geom_rug(alpha = 0.4) +
  labs(title = "Distribution of IDR, based on m6A")

ggplot(r3_idr_m6A, aes(x = ModeScore, group = m6A, color = m6A)) +
  geom_density() +
  geom_rug(alpha = 0.4) +
  labs(title = "Distribution of Mode Scores, based on m6A")

ggplot(r3_idr_m6A, aes(x = T2Cfraction, group = m6A, color = m6A)) +
  geom_density() +
  geom_rug(alpha = 0.4) +
  labs(title = "Distribution of T2C Fraction, based on m6A")

r3_idr_m6A %>%
  filter(T2Cfraction >= 0.25) %>%
  ggplot(aes(x = T2Cfraction, group = m6A, color = m6A)) +
    geom_density() +
    geom_rug(alpha = 0.4) +
    labs(title = "Distribution of T2C Fraction restricted to >= 0.25, based on m6A")


ggplot(r3_idr_m6A, aes(x = ConversionSpecificity, group = m6A, color = m6A)) +
  geom_density() +
  geom_rug(alpha = 0.4) +
  labs(title = "Distribution of Conversion Specificity, based on m6A")

ggplot(r3_idr_m6A, aes(x = log10(T2C_1), group = m6A, color = m6A)) +
  geom_density() +
  geom_rug(alpha = 0.4) +
  labs(title = "Distribution of T2C Count, based on m6A")

```












