---
title: "IDR Testing"
author: "Sarah Arcos"
date: "1/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(tidyverse)
library(RCAS)
library(readxl)
#devtools::install_github('BIMSBbioinfo/RCAS')
#BiocManager::install("org.Hs.eg.db")

library(GenomicFeatures)
library(clusterProfiler)
library(org.Hs.eg.db)
```

#Fix naming of gene_biotype column in gtf rds
```{r}
# gtf <- readRDS("~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.granges.rds")
# 
# gtf$gene_biotype <- gtf$gene_type
# gtf$gene_type <- NULL
# 
# saveRDS(gtf, "~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf.granges.rds")
```


# Read in data
```{r}
r2_bw3 <- readRDS("~/files/PARCLIP_IDR/data/df2_rep2.rda")
r3_bw3 <- readRDS("~/files/PARCLIP_IDR/data/df2_rep3.rda")

#r_fracs <- readRDS("~/files/PARCLIP_IDR/data/r2_r3_overlaps.rda")

#idr_mc <- readRDS("~/files/PARCLIP_IDR/data/idr_summ_T2Cfraction_MC1000")

m6A_r1 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep1.rda")
m6A_r2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep2.rda")
m6A_r1_q0.2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep1_q0.2.rda")
m6A_r2_q0.2 <- readRDS("~/files/PARCLIP_IDR/data/m6A_rep2_q0.2.rda")

#m6A_overlaps <- readRDS("~/files/PARCLIP_IDR/data/m6A_overlaps.rda")

#idr_m6A <- readRDS("~/files/PARCLIP_IDR/data/idr_summ_m6A_MC1000")


#df2_m6A <- readRDS("~/files/PARCLIP_IDR/data/df2_m6A.rda")
```

## Make everything in the same consistent format
 - GRanges
```{r}
df2_rep2 <- GRanges(seqnames = r2_bw3$Chromosome, IRanges(start = r2_bw3$Start, end = r2_bw3$End), strand = r2_bw3$Strand, name = r2_bw3$Name) %>%
  keepStandardChromosomes(pruning.mode="coarse")
df2_rep3 <- GRanges(seqnames = r3_bw3$Chromosome, IRanges(start = r3_bw3$Start, end = r3_bw3$End), strand = r3_bw3$Strand, name = r3_bw3$Name) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep1 <- GRanges(seqnames = m6A_r1$X1, IRanges(start = m6A_r1$X2, end = m6A_r1$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep2 <- GRanges(seqnames = m6A_r2$X1, IRanges(start = m6A_r2$X2, end = m6A_r2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")

df2_ov <- findOverlaps(df2_rep2, df2_rep3, maxgap = 1) %>%
  as_tibble()

df2 <- GRanges(seqnames = r3_bw3$Chromosome[df2_ov$subjectHits], IRanges(start = r3_bw3$Start[df2_ov$subjectHits], end = r3_bw3$End[df2_ov$subjectHits]), strand = r3_bw3$Strand[df2_ov$subjectHits], name = r3_bw3$Name[df2_ov$subjectHits]) %>%
  keepStandardChromosomes(pruning.mode="coarse")

m6A_ov <- findOverlaps(m6A_rep1, m6A_rep2, maxgap = 1) %>%
  as_tibble()

m6A <- GRanges(seqnames = m6A_r1$X1[m6A_ov$queryHits], IRanges(start = m6A_r1$X2[m6A_ov$queryHits], end = m6A_r1$X3[m6A_ov$queryHits])) %>%
  keepStandardChromosomes(pruning.mode="coarse")

```

#Look at low scoring m6A peaks to see if they overlap better
```{r}
m6A_rep1_q <- GRanges(seqnames = m6A_r1_q0.2$X1, IRanges(start = m6A_r1_q0.2$X2, end = m6A_r1_q0.2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")
m6A_rep2_q <- GRanges(seqnames = m6A_r2_q0.2$X1, IRanges(start = m6A_r2_q0.2$X2, end = m6A_r2_q0.2$X3)) %>%
  keepStandardChromosomes(pruning.mode="coarse")

m6A_ov_q <- findOverlaps(m6A_rep1_q, m6A_rep2_q, maxgap = 1) %>%
  as_tibble()

m6A_q <- GRanges(seqnames = m6A_r1_q0.2$X1[m6A_ov_q$queryHits], IRanges(start = m6A_r1_q0.2$X2[m6A_ov_q$queryHits], end = m6A_r1_q0.2$X3[m6A_ov_q$queryHits])) %>%
  keepStandardChromosomes(pruning.mode="coarse")

df2_m6A_q <- subsetByOverlaps(df2, m6A_q) %>%
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_")) %>%
  filter(name != "")

df2 %>% as_tibble() %>% dim()

m6A %>% as_tibble() %>% dim()

m6A_q %>% as_tibble() %>% dim()

df2_m6A %>% as_tibble() %>% dim()

df2_m6A_q %>% as_tibble() %>% dim()



m6A_df2_q <- subsetByOverlaps(m6A_q, df2) %>%
  as_tibble()

m6A_df2 <- subsetByOverlaps(m6A, df2) %>%
  as_tibble()

m6A_df2 %>% as_tibble() %>% dim()

m6A_df2_q %>% as_tibble() %>% dim()

```




#Get tally for each gene of m6A sites, DF2 sites, and shared sites
 - First annotate DF2 sites by presence/absence of m6A
 - Then make dataframe of gene, DF2 site id, m6A site id, and p/a of m6A
 - Then make tallies
 - Then plot distribution

```{r}

df2_df <- df2 %>%
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_")) %>%
  filter(name != "")

df2_m6A <- subsetByOverlaps(df2, m6A) %>%
  as_tibble() %>%
  mutate(peakID = paste(seqnames, start, end, sep = "_")) %>%
  filter(name != "")
  
all_peaks <- df2_df %>%
  mutate(PeakCat = case_when(
    peakID %in% df2_m6A$peakID ~ "m6A_overlapped",
    TRUE ~ "df2_only"
  )) %>%
  group_by(name, PeakCat) %>%
  tally() %>%
  group_by(name) %>%
  mutate(total_peaks = sum(n)) %>%
  spread(key = PeakCat, value = n, fill = 0) %>%
  mutate(df2_only_frac = df2_only / total_peaks)

ggplot(all_peaks, aes(x = df2_only_frac)) +
  geom_bar()

ggplot(all_peaks, aes(x = df2_only)) +
  geom_bar()
```



#Motifs
```{r, eval = FALSE}
df2_df %>%
  filter(peakID %in% df2_m6A$peakID) %>%
  dplyr::select(seqnames, start, end, name, width, strand) %>%
  write_tsv("df2_m6A.bed", col_names = FALSE)

tb <- importBed('df2_m6A.bed')

motifResults <- runMotifDiscovery(queryRegions = tb, 
                           resizeN = 0, sampleN = 10000,
                           genomeVersion = 'hg19', motifWidth = 6,
                           motifN = 4, nCores = 4)

ggseqlogo::ggseqlogo(motifResults$matches_query)

df2_df %>%
  filter(!(peakID %in% df2_m6A$peakID)) %>%
  dplyr::select(seqnames, start, end, name, width, strand) %>%
  write_tsv("df2_no_m6A.bed", col_names = FALSE)

tb_n <- importBed('df2_no_m6A.bed')

motifResults_n <- runMotifDiscovery(queryRegions = tb_n, 
                           resizeN = 0, sampleN = 10000,
                           genomeVersion = 'hg19', motifWidth = 6,
                           motifN = 4, nCores = 4)

ggseqlogo::ggseqlogo(motifResults_n$matches_query)
```

#Metagene
```{r}
gff <- importGtf(filePath = "~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf")
txdbFeatures <- getTxdbFeaturesFromGRanges(gff)

cvgList <- calculateCoverageProfileList(queryRegions = tb, 
                                        targetRegionsList = txdbFeatures, 
                                        sampleN = 10000)

# ggplot2::ggplot(cvgList, aes(x = bins, y = meanCoverage)) + 
#   geom_ribbon(fill = 'lightgreen', 
#               aes(ymin = meanCoverage - standardError * 1.96, 
#                   ymax = meanCoverage + standardError * 1.96)) + 
#   geom_line(color = 'black') + theme_bw(base_size = 14) +
#   facet_wrap( ~ feature, ncol = 3) 

cvgList_tx <- cvgList %>%
  filter(feature  == "transcripts")

ggplot2::ggplot(cvgList_tx, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14)


cvgList_n <- calculateCoverageProfileList(queryRegions = tb_n, 
                                        targetRegionsList = txdbFeatures, 
                                        sampleN = 10000)


cvgList_tx_n <- cvgList_n %>%
  filter(feature  == "transcripts")

ggplot2::ggplot(cvgList_tx_n, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14)

cvgList_all <- bind_rows(list("m6A-modified" = cvgList_tx, "Unmodified" = cvgList_tx_n), .id = "m6A")

ggplot2::ggplot(cvgList_all, aes(x = bins, y = meanCoverage)) + 
  geom_ribbon(fill = 'steelblue', alpha = 0.4,
              aes(ymin = meanCoverage - standardError * 1.96, 
                  ymax = meanCoverage + standardError * 1.96)) + 
  geom_line(color = 'black', size = 0.5) + theme_bw(base_size = 14) +
  facet_wrap( ~ m6A)

```


#Gene regions
```{r}
summary <- summarizeQueryRegions(queryRegions = tb, 
                                 txdbFeatures = txdbFeatures)

df <- data.frame(summary)
df$percent <- round((df$count / length(tb)), 3) * 100
df$feature <- rownames(df)
df <- df %>% filter(feature %in% c("exons", "introns", "threeUTRs",
                                   "fiveUTRs", "cds"))
ggplot2::ggplot(df, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb), ')')) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))

summary_n <- summarizeQueryRegions(queryRegions = tb_n, 
                                 txdbFeatures = txdbFeatures)

df_n <- data.frame(summary_n)
df_n$percent <- round((df_n$count / length(tb_n)), 3) * 100
df_n$feature <- rownames(df_n)
df_n <- df_n %>% filter(feature %in% c("exons", "introns", "threeUTRs",
                                   "fiveUTRs", "cds"))
ggplot2::ggplot(df_n, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_n$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb_n), ')')) +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))


df_all <- bind_rows(list("m6A-modified" = df, "Unmodified" = df_n), .id = "m6A")

ggplot2::ggplot(df_all, aes(x = factor(feature, levels = c("exons", "introns",
                                                           "fiveUTRs", "cds", "threeUTRs")), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  scale_fill_brewer(palette = "Accent") +
  geom_label(aes(y = percent + 3), label = df_all$count) + 
  labs(x = 'transcript feature', y = "percent overlap") +
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none") +
  facet_wrap( ~ m6A)
```


#Gene types
```{r}

overlaps <- as.data.table(queryGff(queryRegions = tb, gffData = gff))

biotype_col <- grep('gene_biotype', colnames(overlaps), value = T)
dfv <- overlaps[,length(unique(queryIndex)), by = biotype_col]
colnames(dfv) <- c("feature", "count")
dfv$percent <- round(dfv$count / length(tb) * 100, 1)
dfv <- dfv[order(count, decreasing = TRUE)]
dfv <- dfv %>% filter(feature %in% c("protein_coding", "miRNA", "lincRNA", "snoRNA", 
                                   "snRNA", "rRNA", "pseudogene"))
ggplot2::ggplot(dfv, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = dfv$count) + 
  scale_fill_brewer(palette = "Accent") +
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb), ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


overlaps_n <- as.data.table(queryGff(queryRegions = tb_n, gffData = gff))

dfv_n <- overlaps_n[,length(unique(queryIndex)), by = biotype_col]
colnames(dfv_n) <- c("feature", "count")
dfv_n$percent <- round(dfv_n$count / length(tb_n) * 100, 1)
dfv_n <- dfv_n[order(count, decreasing = TRUE)]
dfv_n <- dfv_n %>% filter(feature %in% c("protein_coding", "miRNA", "lincRNA", "snoRNA", 
                                   "snRNA", "rRNA", "pseudogene"))
ggplot2::ggplot(dfv_n, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = dfv_n$count) + 
  scale_fill_brewer(palette = "Accent") +
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(tb_n), ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

dfv_all <- bind_rows(list("m6A-modified" = dfv, "Unmodified" = dfv_n), .id = "m6A")


ggplot2::ggplot(dfv_all, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = dfv_all$count) + 
  scale_fill_brewer(palette = "Accent") +
  labs(x = 'gene type', y = 'percent overlap') + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none") +
  facet_wrap( ~ m6A)
```


# Ribosome profiling test
 - Genes with no m6A peaks or DF2 peaks
 - Genes with DF2 peaks but no m6A peaks
 - Genes with both peaks and overlap
 - Genes with both peaks and no overlap
```{r, eval = FALSE}
ribo <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_C-Ribosome_profiling.xlsx", sheet = 3, skip = 3, 
                  col_names = c("Gene", "Ribosome-bound fragments", "Poly(A)+ mRNA input", "Translation Efficiency"))

ggplot(ribo, aes(x  = `Translation Efficiency`)) +
  stat_ecdf()

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(df2_only_cat = case_when(
    df2_only < 6 ~ as.character(df2_only),
    TRUE ~ "> 5"
  )) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(df2_only_cat), group = as.factor(df2_only_cat)))

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(m6A_ov_cat = case_when(
    m6A_overlapped < 6 ~ as.character(m6A_overlapped),
    TRUE ~ "> 5"
  )) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(m6A_ov_cat), group = as.factor(m6A_ov_cat))) 

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(df2_total_cat = case_when(
    total_peaks < 6 ~ as.character(total_peaks),
    TRUE ~ "> 5"
  )) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(df2_total_cat), group = as.factor(df2_total_cat)),
            alpha = 0.6, size = 1) 

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  filter(!is.na(total_peaks)) %>%
  ggplot(aes(y = `Translation Efficiency`, x = as.factor(df2_only))) +
  geom_boxplot()

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  filter(!is.na(total_peaks)) %>%
  ggplot(aes(y = `Translation Efficiency`, x = as.factor(m6A_overlapped))) +
  geom_boxplot()

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  filter(!is.na(total_peaks)) %>%
  ggplot(aes(y = `Translation Efficiency`, x = as.factor(total_peaks))) +
  geom_boxplot()

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  ggplot(aes(y = `Translation Efficiency`, x = df2_only_frac)) +
  geom_point(color = "steelblue", alpha = 0.5)

```

#Halflife data
```{r, warnings = FALSE}
hl_rep1 <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_E-mRNA_Lifetime-rep2.xlsx", sheet = 4, skip = 1) %>%
  dplyr::select(Gene, `log2(siYTHDF2/siControl)`)

ribo <- read_xlsx("~/files/PARCLIP_IDR/data/GSE49339_C-Ribosome_profiling.xlsx", sheet = 3, skip = 3, 
                  col_names = c("Gene", "Ribosome-bound fragments", "Poly(A)+ mRNA input", "Translation Efficiency"))

hl_rep1 %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac == 0 ~ "All Shared",
    df2_only_frac < 1 & df2_only_frac > 0 ~ "Some Shared",
    df2_only_frac == 1 ~ "DF2 Only",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  ggplot(aes(x = `log2(siYTHDF2/siControl)`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5))

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac == 0 ~ "All Shared",
    df2_only_frac < 1 & df2_only_frac > 0 ~ "Some Shared",
    df2_only_frac == 1 ~ "DF2 Only",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5))


ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac < 0.25 ~ " < 25% Free",
    df2_only_frac >= 0.25 & df2_only_frac <= 0.5 ~ "25%-50% Free",
    df2_only_frac > 0.5 & df2_only_frac < 0.75 ~ "50%-75% Free",
    df2_only_frac >= 0.75 & df2_only_frac < 1~ " > 75% Free",
    df2_only_frac == 1 ~ "All Free",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5))
# 
ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac <= 0.5 ~ " < 50% Free",
    df2_only_frac > 0.5 ~ " > 50% Free",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  ggplot(aes(x = `Translation Efficiency`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5))

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  #filter(df2_only_frac < 1) %>%
  mutate(category = case_when(
    df2_only_frac == 0  ~ "All shared",
    df2_only_frac > 0 & df2_only_frac < 0.25 ~ " < 25% Free",
    df2_only_frac >= 0.25 & df2_only_frac <= 0.5 ~ "25%-50% Free",
    df2_only_frac > 0.5 & df2_only_frac < 0.75 ~ "50%-75% Free",
    df2_only_frac >= 0.75  & df2_only_frac < 1 ~ " > 75% Free",
    df2_only_frac == 1 ~ "All free",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  mutate(category = factor(category, levels = c("All shared",
                                                " < 25% Free",
                                                "25%-50% Free",
                                                "50%-75% Free",
                                                " > 75% Free",
                                                "All free",
                                                "Non-target"))) %>%
  ggplot(aes(x = category, y = `Translation Efficiency`)) +
  geom_violin() +
  geom_boxplot(size = 0.5, width = 0.2)

ribo %>%
  left_join(all_peaks, by = c("Gene" = "name")) %>%
  mutate(category = case_when(
    df2_only_frac == 0 ~ "All Shared",
    df2_only_frac < 1 & df2_only_frac > 0 ~ "Some Shared",
    df2_only_frac == 1 ~ "DF2 Only",
    is.na(df2_only_frac) ~ "Non-target"
  )) %>%
  ggplot(aes(x = `Ribosome-bound fragments`)) +
  stat_ecdf(aes(color = as.factor(category), group = as.factor(category))) +
  scale_x_continuous(limits = c(-1.5, 1.5))

```






# runReport test
```{r, eval = FALSE}
tb <- importBed('df2_m6A.bed')
tg <- importGtf('~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf')

test <- queryGff(queryRegions = queryRegions, gff = gff)
overlaps.dt <- data.table(GenomicRanges::as.data.frame(test)) 
backgroundGenes <- unique(gff$gene_id)
targetedGenes <- unique(test$gene_id)

overlaps <- as.data.table(queryGff(queryRegions = tb, gffData = tg))
df <- overlaps[,length(unique(overlappingQuery)), by = gene_type]
colnames(df) <- c("feature", "count")
df$percent <- round(df$count / length(queryRegions) * 100, 1)
df <- df[order(count, decreasing = TRUE)]
ggplot2::ggplot(df, aes(x = reorder(feature, -percent), y = percent)) + 
  geom_bar(stat = 'identity', aes(fill = feature)) + 
  geom_label(aes(y = percent + 0.5), label = df$count) + 
  labs(x = 'transcript feature', y = paste0('percent overlap (n = ', length(queryRegions), ')')) + 
  theme_bw(base_size = 14) + 
  theme(axis.text.x = element_text(angle = 90))


runReport('df2_m6A.bed', '~/files/PARpipe/files/gencode.v19.chr_patch_hapl_scaff.annotation.gtf')
```


#GO analysis
```{r}

gene_cats <- all_peaks %>%
  mutate(category = case_when(
    df2_only_frac == 0 ~ "All Shared",
    df2_only_frac < 1 & df2_only_frac > 0 ~ "Some Shared",
    df2_only_frac == 1 ~ "DF2 Only"
  ))


df2_only_MF <- enrichGO(gene     = gene_cats %>% filter(category == "DF2 Only") %>% pull(name),
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "MF",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable = FALSE)

dotplot(df2_only_MF) +
  scale_color_gradient(trans = "log10", high = "blue", low = "red") +
  labs(title = "GO Molecular Function: Genes with DF2-only peaks only")

shared_MF <- enrichGO(gene     = gene_cats %>% filter(category %in% c("All Shared", "Some Shared")) %>% pull(name),
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "MF",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable = FALSE)

dotplot(shared_MF) +
  scale_color_gradient(trans = "log10", high = "blue", low = "red") +
  labs(title = "GO Molecular Function: Genes with shared peaks")


df2_only_CC <- enrichGO(gene     = gene_cats %>% filter(category == "DF2 Only") %>% pull(name),
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable = FALSE)

dotplot(df2_only_CC) +
  scale_color_gradient(trans = "log10", high = "blue", low = "red") +
  labs(title = "GO Cellular Component: Genes with DF2-only peaks only")

shared_CC <- enrichGO(gene     = gene_cats %>% filter(category %in% c("All Shared", "Some Shared")) %>% pull(name),
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable = FALSE)

dotplot(shared_CC) +
  scale_color_gradient(trans = "log10", high = "blue", low = "red") +
  labs(title = "GO Cellular Component: Genes with shared peaks")

df2_only_BP <- enrichGO(gene     = gene_cats %>% filter(category == "DF2 Only") %>% pull(name),
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable = FALSE)

dotplot(df2_only_BP) +
  scale_color_gradient(trans = "log10", high = "blue", low = "red") +
  labs(title = "GO Biological Process: Genes with DF2-only peaks only")

shared_BP <- enrichGO(gene     = gene_cats %>% filter(category %in% c("All Shared", "Some Shared")) %>% pull(name),
                OrgDb         = org.Hs.eg.db,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable = FALSE)

dotplot(shared_BP) +
  scale_color_gradient(trans = "log10", high = "blue", low = "red") +
  labs(title = "GO Biological Process: Genes with shared peaks")
  
  
# gene_list <- list(
#   shared = gene_cats %>% filter(category %in% c("All Shared", "Some Shared")) %>% pull(name),
#   df2_only = gene_cats %>% filter(category == "DF2 Only") %>% pull(name)
# )
# 
# cat_MF <- compareCluster(gene_list, fun = "enrichGO",
#                            OrgDb = org.Hs.eg.db,
#                            keyType       = 'SYMBOL',
#                            ont           = "MF",
#                            pAdjustMethod = "BH",
#                            pvalueCutoff  = 0.01,
#                            qvalueCutoff  = 0.05,
#                            readable = FALSE)
# dotplot(cat_MF)
# 
# cat_CC <- compareCluster(gene_list, fun = "enrichGO",
#                            OrgDb = org.Hs.eg.db,
#                            keyType       = 'SYMBOL',
#                            ont           = "CC",
#                            pAdjustMethod = "BH",
#                            pvalueCutoff  = 0.01,
#                            qvalueCutoff  = 0.05,
#                            readable = FALSE)
# dotplot(cat_CC, showCategory = 20) +
#   scale_color_gradient(trans = "log10", high = "blue", low = "red")
# 
# cat_BP <- compareCluster(gene_list, fun = "enrichGO",
#                            OrgDb = org.Hs.eg.db,
#                            keyType       = 'SYMBOL',
#                            ont           = "BP",
#                            pAdjustMethod = "BH",
#                            pvalueCutoff  = 0.01,
#                            qvalueCutoff  = 0.05,
#                            readable = FALSE)
# dotplot(cat_BP, showCategory = 20) +
#   scale_color_gradient(trans = "log10", high = "blue", low = "red")

```

```{r}
sessionInfo()
```













