---
title: "T2CFraction_IDR_MC"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(GenomicRanges)
library(idr)
library(tidyverse)
#install.packages("furrr")
library(furrr)
```

#The goal of this markdown is to calculate the IDR for YTHDF2 PAR-CLIP based on T2CFraction. IDR is a rank-based statistic and the T2CFraction has many tied values. This markdown will explore a solution based on adding random noise to the data over 1000 iterations. An attempt will be made to parallelize due to runtime of IDR algorithm (~6 - 10 seconds)

##Read in data and calculate T2CFraction
 - HAVE TO FILTER OUT WHEN T2C == 1, because this is a genomic mutation
```{r}
read_and_calc <- function(infile) {
  read_tsv(infile, col_names = FALSE) %>%
    dplyr::select(X5, "Chromosome" = X7, "Start" = X8, "End" = X9, X11, "Strand" = X12) %>%
    filter(X11 != "-1") %>%
    separate(X5, sep = ",", into = c("Location", "ReadCount", "EndBase")) %>%
    separate(X11, sep = ",", into = c("AlignedTo", "Name", "Gene", "Seq", "Reads", "ModeLocation", 
                                      "ModeScore", "ConLocCount", "ConEvCount", "NonConEvCount", "Annotation")) %>%
    group_by(Gene, Location) %>%
    mutate(TotalReads = sum(as.numeric(ReadCount))) %>%
    dplyr::select(-ReadCount, -EndBase) %>%
    unique() %>%
    spread(Location, TotalReads, fill = 0) %>%
    mutate(T2Cfraction = T2C_1 / (T2C_1 + Other_1 + None),
          ConversionSpecificity = log10(T2C_1 / (Other_1 + 1)),
          ModeScore = as.numeric(ModeScore)) %>%
    filter(T2Cfraction < 1)
}

r2_bw3 <- read_and_calc("~/PARpipe/YTHDF2_rep2/SRR944647.allclusters.bed")
r3_bw3 <- read_and_calc("~/PARpipe/YTHDF2_rep3/SRR944648.allclusters.bed")

```

##Find overlapping peaks
```{r}
p1 <- GRanges(r2_bw3$Chromosome, IRanges(r2_bw3$Start, r2_bw3$End), score=r2_bw3$T2Cfraction, name = r2_bw3$Gene) %>%
  keepStandardChromosomes(pruning.mode="coarse")
p2 <- GRanges(r3_bw3$Chromosome, IRanges(r3_bw3$Start, r3_bw3$End), score=r3_bw3$T2Cfraction, name = r3_bw3$Gene) %>%
  keepStandardChromosomes(pruning.mode="coarse")

overlaps <- findOverlaps(p1, p2, maxgap = 1) %>%
  as_tibble()

#CHECK IF PICKING MIN SCORE STILL MAKES SENSE WITH T2Cfraction
multi_overlaps <- overlaps %>%
  mutate(queryScore = p1$score[.$queryHits],
        subjectScore = p2$score[.$subjectHits]) %>%
  group_by(queryHits) %>%
  mutate(subjectMax = min(subjectScore)) %>%
  ungroup() %>%
  group_by(subjectHits) %>%
  mutate(queryMax = min(queryScore)) %>%
  mutate(topHit = case_when(
    queryScore == queryMax & subjectScore == subjectMax ~ TRUE,
    TRUE ~ FALSE
))

overlaps_max <- multi_overlaps %>%
  filter(topHit == TRUE)

print(paste("Number of peaks breaking ties by max value: ", n_distinct(overlaps_max$queryHits)))

#taking IDs from the bigger replicate
d1 <- p1$score[overlaps_max$queryHits]
d2 <- p2$score[overlaps_max$subjectHits]
ids <- p2$name[overlaps_max$subjectHits]

#Pipe function
simplify_counts <- . %>%
  select(name, None, Other_1, T2C_1) %>%
  mutate(No_T2C = None + Other_1) %>%
  select(-None, -Other_1)

p1_counts <- tibble(
  name = p1$name[overlaps_max$queryHits]
) %>%
  left_join(r2_bw3, by = c("name" = "Gene")) %>%
  simplify_counts()

p2_counts <- tibble(
  name = p2$name[overlaps_max$subjectHits]
) %>%
  left_join(r3_bw3, by = c("name" = "Gene")) %>%
  simplify_counts()

#Break ties manually by randomly shuffling the order of tied scores. This is what get's randomized and run 1000 times

```


##Function to calculate IDR
```{r}

add_noise <- function(count_df) {
  count_df %>%
    mutate()
}

#using variables defined above rather than inside function to simplify furrr call below
idr_calc <- function(n){
  #IDR calculation
  system.time({
    res <- est.IDR(cbind(jitter(d1), jitter(d2)), mu=3, sigma=1, rho=.9, p=.5)
  })
# 
#   df <- data.frame(rep1=dat[,1],rep2=dat[,2],
#                    rank1=rank(-dat[,1]),rank2=rank(-dat[,2]),
#                    id = ids,
#                    idr=res$idr)

  return(data.frame(id = ids,
                   idr = res$idr))
}

test <- idr_calc(1)

test %>%
    mutate(met_cutoff = case_when(
      idr < 0.25 ~ TRUE,
      TRUE ~ FALSE
    )) %>%
    ggplot(aes(rank1, rank2, col = met_cutoff)) +
    scale_color_brewer(palette="Set1") +
    geom_point(size = 0.1, alpha = 0.5)+
    labs(title = paste("Points colored by IDR < 0.25"))

test %>%
    mutate(met_cutoff = case_when(
      idr < 0.25 ~ TRUE,
      TRUE ~ FALSE
    )) %>%
    ggplot(aes(rep1, rep2, col = met_cutoff)) +
    scale_color_brewer(palette="Set1") +
    geom_point(size = 0.1, alpha = 0.5)+
    labs(title = paste("Points colored by IDR < 0.25"))

```

##Iterate using furrr
```{r}
plan(multiprocess)

v <- 1:1000
names(v) = 1000

start_f <- Sys.time()
t <- future_map_dfr(v, idr_calc, .id = "Iteration")
end_f <- Sys.time()

end_f - start_f

saveRDS(t, file = "idr_T2Cfraction_MC1000")
```

##Look at variance of idrs per gene
```{r}
n_distinct(t$id)

t_summary <- t %>%
  group_by(id) %>%
  summarize(v = var(idr),
            sd = sd(idr),
            av = mean(idr))

n_distinct(t_summary %>%
             filter(av < 0.25))
```

##Trying a weird plotting idea
 - Plot the 95% confidence interval:
  - Take the quantile of .025 and .975 and plot that range (ymin = q(0.25), ymax = q(.975))
```{r}
t_summary %>%
  ggplot(aes(x = reorder(id, av), y = av)) +
    #geom_point(size = 0.001, alpha = 0.01) +
    geom_errorbar(aes(ymin = av-sd, ymax = av+sd), color = "steelblue", alpha = 0.5) +
    geom_hline(yintercept = 0.25, color = "red", size = 0.1, alpha = 0.5) +
    scale_y_log10() +
    theme_minimal() +
    theme(axis.text.x = element_blank())


```

## Filter for values that have weirdly large SD, then plot points
```{r}
t_summary %>%
  mutate(weird = case_when(
    sd > av & av-sd < 0.25 ~ "Weird",
    sd > av & av-sd >= 0.25 ~ "IM A PROBLEM",
    TRUE ~ "Normal"
  )) %>%
  ggplot(aes(x = reorder(id, av), y = av, color = weird)) +
    geom_point(size = 0.1, alpha = 0.5) +
    #geom_errorbar(aes(ymin = av-sd, ymax = av+sd), color = "steelblue", alpha = 0.5) +
    geom_hline(yintercept = 0.25, color = "red", size = 0.1, alpha = 0.5) +
    scale_y_log10() +
    theme_minimal() +
    theme(axis.text.x = element_blank()) +
    scale_color_manual(values=c("grey", "blue", "red"))
```

##Look at highly tied values
```{r}
test %>%
  group_by(rep1) %>%
  filter(n_distinct > 1) %>%
  ggplot(aes(rep1)) +
  geom_bar()


test <- tibble(d = d1) %>% 
  count(d) %>%
  filter(n > 1) 

test %>%
  ggplot(aes(x = d, y = n)) +
  geom_col()
```







