---
title: "T2CFraction_IDR_MC"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(GenomicRanges)
library(idr)
#library(tidyverse)
#install.packages("furrr")
library(furrr)
library(tidyr)
library(readr)
library(dplyr)
library(ggplot2)
```

#The goal of this markdown is to calculate the IDR for YTHDF2 PAR-CLIP based on T2CFraction. IDR is a rank-based statistic and the T2CFraction has many tied values. This markdown will explore a solution based on adding random noise to the data over 1000 iterations. An attempt will be made to parallelize due to runtime of IDR algorithm (~6 - 10 seconds)

##Read in data and calculate T2CFraction
 - HAVE TO FILTER OUT WHEN T2C == 1, because this is a genomic mutation
```{r}
read_and_calc <- function(infile) {
  read_tsv(infile, col_names = FALSE) %>%
    dplyr::select(X5, "Chromosome" = X7, "Start" = X8, "End" = X9, X11, "Strand" = X12) %>%
    filter(X11 != "-1") %>%
    separate(X5, sep = ",", into = c("Location", "ReadCount", "EndBase")) %>%
    separate(X11, sep = ",", into = c("AlignedTo", "Name", "Gene", "Seq", "Reads", "ModeLocation", 
                                      "ModeScore", "ConLocCount", "ConEvCount", "NonConEvCount", "Annotation")) %>%
    group_by(Gene, Location) %>%
    mutate(TotalReads = sum(as.numeric(ReadCount))) %>%
    dplyr::select(-ReadCount, -EndBase) %>%
    unique() %>%
    spread(Location, TotalReads, fill = 0) %>%
    mutate(T2Cfraction = T2C_1 / (T2C_1 + Other_1 + None),
          ConversionSpecificity = log10(T2C_1 / (Other_1 + 1)),
          ModeScore = as.numeric(ModeScore)) %>%
    filter(T2Cfraction < 1)
}

#Uncomment to re-read files

# r2_bw3 <- read_and_calc("~/PARpipe/YTHDF2_rep2/SRR944647.allclusters.bed")
# r3_bw3 <- read_and_calc("~/PARpipe/YTHDF2_rep3/SRR944648.allclusters.bed")
# 
# saveRDS(r2_bw3, "~/PARCLIP_IDR/data/df2_rep2.rda")
# saveRDS(r3_bw3, "~/PARCLIP_IDR/data/df2_rep3.rda")

#For using existing RDAs
r2_bw3 <- readRDS("~/PARCLIP_IDR/data/df2_rep2.rda")
r3_bw3 <- readRDS("~/PARCLIP_IDR/data/df2_rep3.rda")

```

##Find overlapping peaks

```{r}
#Run this line to use previously calculated data
# r_fracs <- readRDS("~/PARCLIP_IDR/data/r2_r3_overalps.rda")

#Turn into GRanges object
r2 <- GRanges(r2_bw3$Chromosome, IRanges(r2_bw3$Start, r2_bw3$End), score=r2_bw3$T2Cfraction, name = r2_bw3$Gene) %>%
  keepStandardChromosomes(pruning.mode="coarse")
r3 <- GRanges(r3_bw3$Chromosome, IRanges(r3_bw3$Start, r3_bw3$End), score=r3_bw3$T2Cfraction, name = r3_bw3$Gene) %>%
  keepStandardChromosomes(pruning.mode="coarse")

overlaps <- findOverlaps(r2, r3, maxgap = 1) %>%
  as_tibble()

#For multi overlaps, pick the peaks with the maximum score
multi_overlaps <- overlaps %>%
  mutate(queryScore = r2$score[.$queryHits],
        subjectScore = r3$score[.$subjectHits]) %>%
  group_by(queryHits) %>%
  mutate(subjectMax = max(subjectScore)) %>%
  ungroup() %>%
  group_by(subjectHits) %>%
  mutate(queryMax = max(queryScore)) %>%
  mutate(topHit = case_when(
    queryScore == queryMax & subjectScore == subjectMax ~ TRUE,
    TRUE ~ FALSE
))

overlaps_max <- multi_overlaps %>%
  filter(topHit == TRUE)

print(paste("Number of peaks breaking ties by max value: ", n_distinct(overlaps_max$queryHits)))

r2_fracs <- tibble(
  r2_name = r2$name[overlaps_max$queryHits],
  r2_score = r2$score[overlaps_max$queryHits]
)
r3_fracs <- tibble(
  r3_name = r3$name[overlaps_max$subjectHits],
  r3_score = r3$score[overlaps_max$subjectHits]
)

#make one big 
r_fracs <- bind_cols(r2_fracs, r3_fracs)

saveRDS(r_fracs, "~/PARCLIP_IDR/data/r2_r3_overlaps.rda")

```


##Function to calculate IDR
 - r rank function ranks low values as lower ranks, with high values as high ranks. (ie, rank 1 might be 0.001 and rank 100 might be .90). So if IDR function is expecting high confidence to mean high score, then we should not reverse ranks if we are passing ranks directly to idr. Because in this case rank == score, so rank 100 for 0.90 is what we are going for.
 - shown below, using raw ranks or score as input to idr function gives the exact same distribution of resulting idrs (as long as jitter is used to break ties in raw score)
 - conclusion: calculate idr using ranks as input, with ties broken randomly. This avoids possibility of jitter causing scores to flip rank that should not flip.
```{r}
#using variables defined above rather than inside function to simplify furrr call below
idr_calc <- function(score_df){
  #IDR calculation
    res <- est.IDR(cbind(rank(score_df$r2_score, ties.method = "random"), 
                         rank(score_df$r3_score, ties.method = "random")), 
                         mu=3, sigma=1, rho=.9, p=.5)

  return(bind_cols(score_df,
                   tibble(idr = res$IDR)))
}

test <- idr_calc_rank(r_fracs)
test2 <- idr_calc_rank(r_fracs)

test %>%
  ggplot(aes(x = r3_score, y = r2_score, color = idr)) +
  geom_point(alpha = 0.5, size = 1)

#keeping in mind the ranking as described above
test %>%
    ggplot(aes(rank(r2_score, ties.method = "random"), rank(r3_score, ties.method = "random"), color = idr)) +
    geom_point(size = 0.1, alpha = 0.5)

test %>%
  ggplot(aes(x = r3_score, y = r2_score, color = idr)) +
  geom_point(alpha = 0.5, size = 1)

test %>%
    ggplot(aes(rank(-r2_score, ties.method = "random"),r2_score, color = idr)) +
    geom_point(size = 1, alpha = 0.5)

#Compare rank vs. score idr distributions to ensure they are very similar
idr_calc_score <- function(score_df){
  #IDR calculation
    res <- est.IDR(cbind(jitter(score_df$r2_score), jitter(score_df$r3_score)), 
                         mu=3, sigma=1, rho=.9, p=.5)

  return(bind_cols(score_df,
                   tibble(idr = res$IDR)))
}

test_s <- idr_calc_score(r_fracs)

tibble(r1 = test$idr,
       r2 = test2$idr,
       s = test_s$idr) %>%
  gather(dataset, idr) %>%
  ggplot(aes(x = log10(idr), color = dataset, group = dataset)) +
  geom_density()
  
```

##Iterate using furrr
 - Takes ~ 25 minutes for 1000 trials
```{r}
idr_calc_parallel <- function(n, score_df){
  #IDR calculation
    res <- est.IDR(cbind(rank(score_df$r2_score, ties.method = "random"), 
                         rank(score_df$r3_score, ties.method = "random")), 
                         mu=3, sigma=1, rho=.9, p=.5)

  return(bind_cols(score_df,
                   tibble(idr = res$IDR)) %>%
           select(r2_name, r3_name, idr))
}

plan(multiprocess)

v <- 1:1000
names(v) = 1000

start_f <- Sys.time()
t <- future_map_dfr(v, idr_calc_parallel, r_fracs, .id = "Iteration")
end_f <- Sys.time()

end_f - start_f

saveRDS(t, file = "~/PARCLIP_IDR/data/idr_T2Cfraction_MC1000")
```

##Get summary statistics of idrs per peak and plot distribution of mean(idr)
```{r}
t_summary <- t %>%
  select(-r2_name) %>%
  group_by(r3_name) %>%
  summarize(v = var(idr),
            sd = sd(idr),
            av = mean(idr),
            ci_lower = quantile(idr, 0.025),
            ci_upper = quantile(idr, 0.975))

saveRDS(t_summary, "~/PARCLIP_IDR/data/idr_summ_T2Cfraction_MC1000")

ggplot(t_summary, aes(x = av)) +
  geom_density()
```

##Plot the 95% confidence interval:
  - Take the quantile of .025 and .975 and plot that range
```{r}
t_summary %>%
  ggplot(aes(x = reorder(r3_name, av), y = av)) +
    #geom_point(size = 0.001, alpha = 0.01) +
    geom_errorbar(aes(ymin = ci_lower, ymax = ci_upper), color = "steelblue", alpha = 0.5) +
    geom_hline(yintercept = 0.25, color = "red", size = 0.1, alpha = 0.5) +
    #scale_y_log10() +
    theme_minimal() +
    theme(axis.text.x = element_blank()) +
    labs(x = "Peaks, ordered by mean IDR", y = "Mean IDR", title = "95% Confidence Interval of IDR Values from 1000 MC Trials")


```









